File: runtimes/compose/src/main/java/com/styleconverter/runtime/columns/VerticalFragmentGeometry.kt
// Wave-47 lane Z2 — pure fragment geometry for css-break-3 §4 multicol
// fragmentation under VERTICAL writing modes. No Compose imports on purpose:
// the JUnit suite pins this math on the plain JVM, and the identical V-table
// is pinned on the iOS runtime (VerticalFragmentGeometry.swift) so both
// platforms fragment byte-identically — the same cross-platform contract
// FragmentGeometry.kt documents for horizontal-tb.
package com.styleconverter.runtime.columns

/**
 * Fragment list for a vertical-writing-mode multicol container's sole
 * over-wide child, per css-multicol-1 §2 + css-break-3 §4 with the
 * `box-decoration-break: slice` default of §6:
 *
 * Axes transpose against [FragmentGeometry]: the container's INLINE axis is
 * vertical, so the N column boxes STACK VERTICALLY, each `colH` tall
 * (`colH = (availableInline − (N−1)·G) / N`) and W wide, where W — the
 * container's content-box WIDTH — is the fragmentainer BLOCK size the child
 * breaks at. The child lays out ONCE as a continuous box of block extent C
 * (its used physical WIDTH) × colH; fragment i then shows one W-wide band of
 * that continuous paint inside column i — a pure clip + translate, consumed
 * by the same drawWithContent replay as the horizontal pass.
 *
 * The BLOCK-FLOW DIRECTION decides which physical band fragment i shows and
 * where a partial band sits (css-writing-modes-4 §6.4):
 *  * vertical-lr — block flows left→right: fragment i is the child's
 *    physical x-band [i·W, (i+1)·W), left-aligned in its column
 *    (translateX = −i·W); the last partial band leaves the column's RIGHT
 *    remainder showing the container's own paint.
 *  * vertical-rl — block flows right→left: fragment i is the band
 *    [C−(i+1)·W, C−i·W), RIGHT-aligned in its column. One formula covers
 *    full and partial bands: translateX = W − C + i·W (band start lands at
 *    max(0, W − remaining), pinned by the V-table test against the
 *    css-break background-image-001 / borders-007 reference geometry).
 */
object VerticalFragmentGeometry {

    /**
     * Build the fragment list.
     *
     * @param childBlockSizePx  C — the child's laid-out physical width.
     * @param columnBlockSizePx W — the container's content width (the
     *        fragmentainer block size each fragment breaks at).
     * @param columnInlineSizePx colH — one column box's height
     *        ((availableInline − gaps) / N, resolved by the caller).
     * @param columnGapPx       G — the used column-gap (between columns,
     *        i.e. VERTICALLY here).
     * @param columnCount       N — the used column count (caps F, the
     *        column-fill:auto overflow clip — same rule as horizontal).
     * @param blockRtl          true for vertical-rl / sideways-rl (block
     *        axis right→left), false for vertical-lr (left→right).
     */
    fun fragmentGeometry(
        childBlockSizePx: Int,
        columnBlockSizePx: Int,
        columnInlineSizePx: Int,
        columnGapPx: Int,
        columnCount: Int,
        blockRtl: Boolean,
    ): List<FragmentGeometry.Fragment> {
        // Defensive floors — the same contracts as FragmentGeometry: used
        // count is a positive <integer> (css-multicol §3.2), gaps are
        // non-negative (css-align §8), negative extents clip to nothing.
        val count = maxOf(1, columnCount)
        val gap = maxOf(0, columnGapPx)
        val colH = maxOf(0, columnInlineSizePx)
        val w = columnBlockSizePx
        // No fragmentainer extent → the degenerate single fragment (empty
        // clip), mirroring the horizontal guard: callers gate on C > W > 0
        // and never reach this branch in practice.
        if (w <= 0) {
            return listOf(FragmentGeometry.Fragment(0, 0, 0, maxOf(0, w), colH, 0, 0))
        }
        val c = maxOf(0, childBlockSizePx)
        // F = ceil(C / W) in Long space (overflow-safe), at least 1, capped
        // at N — the paint past column N−1 simply never gets a column, which
        // IS the column-fill:auto overflow clip (same rule as horizontal).
        val rawCount = (c.toLong() + w - 1) / w
        val fragmentCount = rawCount.coerceIn(1L, count.toLong()).toInt()
        return List(fragmentCount) { i ->
            // Column i's vertical origin — columns stack downward with
            // exactly i gaps above column i (css-multicol §3 box placement,
            // transposed to the vertical inline axis).
            val y = i * (colH + gap)
            // Block-band translation (doc comment above): left-aligned
            // −i·W bands for lr; right-aligned W−C+i·W bands for rl.
            val tx = if (blockRtl) w - c + i * w else -i * w
            // Clip = column i's rect in container coordinates (full content
            // width × colH); translate moves band i of the continuous child
            // paint into it (+y vertically — the child is laid out at the
            // container origin, colH tall).
            FragmentGeometry.Fragment(
                columnIndex = i,
                clipLeft = 0,
                clipTop = y,
                clipWidth = w,
                clipHeight = colH,
                translateX = tx,
                translateY = y,
            )
        }
    }

    /**
     * The used column inline size colH = (availableInline − (N−1)·G) / N,
     * floored at 0 — the css-multicol §3.4 width fit transposed to the
     * vertical inline axis. Shared by the measure helper and the V-table
     * test so the geometry and the gate can never disagree.
     */
    fun columnInlineSizePx(availableInlinePx: Int, columnCount: Int, columnGapPx: Int): Int {
        val count = maxOf(1, columnCount)
        val gap = maxOf(0, columnGapPx)
        return maxOf(0, (availableInlinePx - (count - 1) * gap) / count)
    }
}

File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/columns/VerticalFragmentGeometry.swift
//
//  VerticalFragmentGeometry.swift
//  StyleEngine/columns — wave 47 (lane Z2).
//
//  Pure fragment geometry for css-break-3 §4 multicol fragmentation under
//  VERTICAL writing modes — the iOS half of the shared V-table (the Compose
//  twin is columns/VerticalFragmentGeometry.kt; VerticalFragmentGeometryTests
//  pins the identical expected values, so drift is a cross-platform
//  divergence, not a refactor).
//
//  Axes transpose against FragmentGeometry: the container's INLINE axis is
//  vertical, so the N column boxes STACK VERTICALLY, each colH tall
//  (colH = (availableInline − (N−1)·G) / N) and W wide, where W — the
//  container's content-box WIDTH — is the fragmentainer BLOCK size the child
//  breaks at. The child lays out ONCE as a continuous box of block extent C
//  (its used physical WIDTH) × colH; fragment i shows one W-wide physical
//  band of that continuous paint inside column i.
//
//  The BLOCK-FLOW DIRECTION decides which band fragment i shows and where a
//  partial band sits (css-writing-modes-4 §6.4):
//   * vertical-lr — bands walk rightward: [i·W, (i+1)·W), left-aligned
//     (translate x = −i·W);
//   * vertical-rl — bands walk leftward: [C−(i+1)·W, C−i·W), RIGHT-aligned;
//     one formula covers full and partial bands: translate x = W − C + i·W.
//

// CoreGraphics for CGRect/CGSize (the Fragment shape); Foundation for ceil.
import CoreGraphics
import Foundation

// Namespacing enum — all members static + pure (never instantiated).
enum VerticalFragmentGeometry {

    /// The vertical V-table. Reuses FragmentGeometry.Fragment: clipRect is
    /// column i's rect in container content coordinates — (0, y_i, W, colH)
    /// with y_i = i·(colH+G) — and translate carries (tx_i, +y_i). The
    /// renderer's VStack consumption supplies the vertical slot position
    /// itself and reads only translate.width, exactly like the horizontal
    /// HStack ignores translate.width and reads .height.
    ///
    /// UNLIKE the horizontal S-table, the FITS case (C ≤ W) still returns a
    /// single fragment: under vertical-rl even fitting content is
    /// RIGHT-aligned in its column (block-start is the right edge), which
    /// the unfragmented path cannot express — the Compose V5 row pins the
    /// same choice.
    ///
    /// - Parameters:
    ///   - childBlockSize: C — the child's block extent (used width) in px.
    ///   - containerBlockSize: W — the container's content-box width, the
    ///     fragmentainer block size (non-positive → nil, identity path).
    ///   - columnInlineSize: colH — one column box's height in px.
    ///   - gapPx: G — the used column-gap (vertical here; floored at 0).
    ///   - columnCount: N — the §3.4 used count (caps F — the
    ///     column-fill:auto overflow clip, same rule as horizontal).
    ///   - blockRtl: true for vertical-rl / sideways-rl (bands walk left).
    static func fragments(childBlockSize: Double,
                          containerBlockSize: Double,
                          columnInlineSize: Double,
                          gapPx: Double,
                          columnCount: Int,
                          blockRtl: Bool) -> [FragmentGeometry.Fragment]? {
        // No fragmentainer extent → nothing to slice at (identity path).
        guard containerBlockSize > 0 else { return nil }
        // Defensive floors mirroring the horizontal table's invariants.
        let w = containerBlockSize
        let colH = max(0, columnInlineSize)
        let g = max(0, gapPx)
        let n = max(1, columnCount)
        let c = max(0, childBlockSize)
        // F = ceil(C/W), at least 1 (the fits case — see the doc above),
        // capped at N (the overflow tail never gets a column).
        let count = min(n, max(1, Int(ceil(c / w))))
        return (0..<count).map { i in
            // Column i's vertical origin — exactly i gaps above it.
            let y = Double(i) * (colH + g)
            // Band translation (doc above): left-aligned −i·W for lr,
            // right-aligned W−C+i·W for rl.
            let tx = blockRtl ? w - c + Double(i) * w : -Double(i) * w
            return FragmentGeometry.Fragment(
                columnIndex: i,
                // The clip is column i's rect — full content width × colH.
                clipRect: CGRect(x: 0, y: y, width: w, height: colH),
                // (tx, +y): the y rides for container-coordinate parity
                // with the Compose draw replay; the VStack consumer
                // supplies the slot position and reads only tx.
                translate: CGSize(width: tx, height: y))
        }
    }

    /// colH = (availableInline − (N−1)·G) / N floored at 0 — the §3.4 fit
    /// transposed to the vertical inline axis (Compose twin:
    /// VerticalFragmentGeometry.columnInlineSizePx).
    static func columnInlineSize(availableInline: Double,
                                 columnCount: Int,
                                 gapPx: Double) -> Double {
        let n = max(1, columnCount)
        let g = max(0, gapPx)
        return max(0, (availableInline - Double(n - 1) * g) / Double(n))
    }
}
